packages/SwingSet/src/kernel/collectionManager.js

```diff
@@ -0,0 +1,891 @@
+}
+
+const BIGINT_TAG_LEN = 10;
+const BIGINT_LEN_MODULUS = 10 ** BIGINT_TAG_LEN;
+
+function bigintToDBEntryKey(n) {
+  if (n < 0n) {
+    const raw = (-n).toString();
+    const modulus = 10n ** BigInt(raw.length);
+    const numstr = (modulus + n).toString(); // + because n is negative
+    const lenTag = zeroPad(BIGINT_LEN_MODULUS - raw.length, BIGINT_TAG_LEN);
+    return `n${lenTag}:${zeroPad(numstr, raw.length)}`;
+  } else {
+    const numstr = n.toString();
+    return `p${zeroPad(numstr.length, BIGINT_TAG_LEN)}:${numstr}`;
+  }
+}
+
+function dbEntryKeyToBigint(k) {
+  const numstr = k.substring(BIGINT_TAG_LEN + 2);
+  const n = BigInt(numstr);
+  if (k[0] === 'n') {
+    const modulus = 10n ** BigInt(numstr.length);
+    return -(modulus - n);
+  } else {
+    return n;
+  }
+}
+
+function pattEq(p1, p2) {
+  return compareRank(p1, p2) === 0;
+}
+
+export function makeCollectionManager(
+  syscall,
+  vrm,
+  allocateExportID,
+  getSlotForVal,
+  getValForSlot,
+  registerEntry,
+  serialize,
+  unserialize,
+) {
+  const storeKindIDToName = new Map();
+
+  let storeKindInfoNeedsInitialization = true;
+  const storeKindInfo = {
+    scalarMapStore: {
+      hasWeakKeys: false,
+      kindID: 0,
+      // eslint-disable-next-line no-use-before-define
+      reanimator: reanimateScalarMapStore,
+      durable: false,
+    },
+    scalarWeakMapStore: {
+      hasWeakKeys: true,
+      kindID: 0,
+      // eslint-disable-next-line no-use-before-define
+      reanimator: reanimateScalarWeakMapStore,
+      durable: false,
+    },
+    scalarSetStore: {
+      hasWeakKeys: false,
+      kindID: 0,
+      // eslint-disable-next-line no-use-before-define
+      reanimator: reanimateScalarSetStore,
+      durable: false,
+    },
+    scalarWeakSetStore: {
+      hasWeakKeys: true,
+      kindID: 0,
+      // eslint-disable-next-line no-use-before-define
+      reanimator: reanimateScalarWeakSetStore,
+      durable: false,
+    },
+    scalarDurableMapStore: {
+      hasWeakKeys: false,
+      kindID: 0,
+      // eslint-disable-next-line no-use-before-define
+      reanimator: reanimateScalarMapStore,
+      durable: true,
+    },
+    scalarDurableWeakMapStore: {
+      hasWeakKeys: true,
+      kindID: 0,
+      // eslint-disable-next-line no-use-before-define
+      reanimator: reanimateScalarWeakMapStore,
+      durable: true,
+    },
+    scalarDurableSetStore: {
+      hasWeakKeys: false,
+      kindID: 0,
+      // eslint-disable-next-line no-use-before-define
+      reanimator: reanimateScalarSetStore,
+      durable: true,
+    },
+    scalarDurableWeakSetStore: {
+      hasWeakKeys: true,
+      kindID: 0,
+      // eslint-disable-next-line no-use-before-define
+      reanimator: reanimateScalarWeakSetStore,
+      durable: true,
+    },
+  };
+
+  function prefixc(collectionID, dbEntryKey) {
+    return `vc.${collectionID}.${dbEntryKey}`;
+  }
+
+  function obtainStoreKindID(kindName) {
+    if (storeKindInfoNeedsInitialization) {
```

Comment:
The kind IDs assigned to the built-in collections will also vary, depending upon how many `makeKind()` calls were made by userspace before they created their first virtual collection. The code tracks those IDs correctly, in a vatstore key named `storeKindIDTable`, it's just that the IDs themselves will differ from one vat to another. Not fatal, but consistency would be an argument in favor of doing the allocation ahead of time. OTOH if we upgrade to a newer liveslots as part of the vat upgrade, and that newer liveslots has more collection types, their IDs will necessarily be scattered about anyways, because userspace Kinds have already claimed some range.

So, let's leave the allocation to happen on-demand, but let's make sure it handles errors in `syscall.vatstoreSet` properly, knowing that `obtainStoreKindID` might be called by userspace during module initialization, which might catch any error we return. That means any attempt to make a syscall during module initialization must not only fail, but it must also kill the vat, even if the error it raises gets caught.

Reply:
Deferring allocation of the kind ID was initially a workaround for the unavailability of syscalls during vat initialization, but I kept it for both reasons you've already identified: avoiding generating IDs in the common case that they're not needed and dealing with the appearance of new store types later in the vat's life.  I think those are both good reasons for retaining the quasi-lazy approach.  I do not believe the failure scenario you outline can actually happen the way you posit once vatstore operations are allowed during startup.